apps/nsqd/main.go

```diff
@@ -25,7 +28,8 @@ type program struct {
 
 func main() {
 	prg := &program{}
-	if err := svc.Run(prg, syscall.SIGINT, syscall.SIGTERM); err != nil {
+	// SIGTERM handling is conditional on config settings so it's initialized in Start()
+	if err := svc.Run(prg, syscall.SIGINT); err != nil {
```

Comment:
The other issue (constraint?) is that `go-svc` is a single shot signal handling, so if it handles a `SIGTERM` it's done - it won't be able to handle a `SIGINT` later (which we want). I think we want to still keep that package for what it provides us on windows so this (narrowing it's handling to `SIGINT`) seemed like a good middle ground.

I think having signal handling in `main()` is still a good separation point i think it makes the `nsqd` package more testable.

Given those additional details, are there specific ideas you have to refactor this?

Reply:
> The other issue (constraint?) is that go-svc is a single shot signal handling, so if it handles a SIGTERM it's done - it won't be able to handle a SIGINT later (which we want).

Maybe `go-svc` should be enhanced to have some sort of `Signal(int) bool` callback where the application can act on specific signals that the service receives. It could indicate whether to "shutdown" to `go-svc` based on the return code from the callback.

> I think having signal handling in main() is still a good separation point i think it makes the nsqd package more testable.

Yes, signal handling should be in "main" (delegated to `go-svc`), but the logic on lines [https://github.com/nsqio/nsq/pull/1305/files#diff-8a637790ff937071a9d12c7676c7cc9dc0cb8ff3718f1e147dfa117f03862c79R96-R99](96-99) feels like it should be in `nsqd` package (note there's no other business logic that acts on anything in `opts` in `main.go`).

Reply:
Updated. I see which part you are describing now.